src/multi/intersection.rs

```diff
@@ -49,10 +49,14 @@ enum Equality<'a, T: 'a> {
 fn test_equality<'a, T: Ord>(slices: &[&'a [T]]) -> Equality<'a, T> {
     let mut is_equal = true;
     let mut max = &slices[0][0];
-    for x in slices {
-        let x = &x[0];
-        if is_equal { is_equal = max == x }
-        max = cmp::max(max, x);
+    for s in slices {
+        let x = &s[0];
+        if x != max {
+            is_equal = false;
+            if x > max {
+                max = x;
+            }
+        }
```

Comment:
Are you sure it's faster or even simpler to read?

Reply:
By bookending the section with hardware breakpoints, then debugging the unit tests with --profile=release, I viewed the generated disassembly. The original does 2 separate comparison groups (cmpl -> sete + cmovleq) and (testb -> cmovnel). The new does 1 comparison and all 3 conditional moves off that (cmpl -> cmovleq + cmovneq + cmovnel), and is 2 less instructions overall.

I personally find the new syntax a bit easier to read. If you don't like it, though, I can roll this back to the old version. It doesn't make a big difference I think.

Source for reference:
```rust
// Original implementation:
#[inline]
fn test_equality<'a, T: Ord>(slices: &[&'a [T]]) -> Equality<'a, T> {
    let mut is_equal = true;
    let mut max = &slices[0][0];
    unsafe {std::arch::asm!("int3");} // hardware breakpoint and marker
    for x in slices {
        let x = &x[0];
        if is_equal { is_equal = max == x }
        max = cmp::max(max, x);
    }
    unsafe {std::arch::asm!("int3");} // hardware breakpoint and marker
    if is_equal { Equal(max) } else { NotEqual(max) }
}
```
```rust
// New implementation:
#[inline]
fn test_equality<'a, T: Ord>(slices: &[&'a [T]]) -> Equality<'a, T> {
    let mut is_equal = true;
    let mut max = &slices[0][0];
    unsafe {std::arch::asm!("int3");} // hardware breakpoint and marker
    for s in slices {
        let x = &s[0];
        if x != max {
            is_equal = false;
            if x > max {
                max = x;
            }
        }
    }
    unsafe {std::arch::asm!("int3");} // hardware breakpoint and marker
    if is_equal { Equal(max) } else { NotEqual(max) }
}
```

```assembly
// Original assembly between "int3" markers
5555555A7E2E: CC                         int3   
5555555A7E2F: B0 01                      movb   $0x1, %al
5555555A7E31: 31 C9                      xorl   %ecx, %ecx
5555555A7E33: 66 2E 0F 1F 84 00 00 00 >  nopw   %cs:(%rax,%rax)
5555555A7E3D: 0F 1F 00                   nopl   (%rax)
5555555A7E40: 49 83 7C 0E 08 00          cmpq   $0x0, 0x8(%r14,%rcx)  ; loop start
5555555A7E46: 0F 84 52 02 00 00          je     0x5555555a809e  ; <+782>
5555555A7E4C: 49 8B 34 0E                movq   (%r14,%rcx), %rsi
5555555A7E50: 8B 3A                      movl   (%rdx), %edi
5555555A7E52: 31 ED                      xorl   %ebp, %ebp
5555555A7E54: 3B 3E                      cmpl   (%rsi), %edi
5555555A7E56: 40 0F 94 C5                sete   %bpl
5555555A7E5A: 48 0F 4E D6                cmovleq %rsi, %rdx
5555555A7E5E: A8 01                      testb  $0x1, %al
5555555A7E60: 0F B6 C0                   movzbl %al, %eax
5555555A7E63: 0F 45 C5                   cmovnel %ebp, %eax
5555555A7E66: 48 83 C1 10                addq   $0x10, %rcx
5555555A7E6A: 49 39 CC                   cmpq   %rcx, %r12
5555555A7E6D: 75 D1                      jne    0x5555555a7e40  ; <+176>  loop end
5555555A7E6F: CC                         int3   
```
```assembly
// New assembly between "int3" markers
5555555A7E3E: CC                         int3   
5555555A7E3F: B0 01                      movb   $0x1, %al
5555555A7E41: 31 C9                      xorl   %ecx, %ecx
5555555A7E43: 66 2E 0F 1F 84 00 00 00 >  nopw   %cs:(%rax,%rax)
5555555A7E4D: 0F 1F 00                   nopl   (%rax)
5555555A7E50: 49 83 7C 0E 08 00          cmpq   $0x0, 0x8(%r14,%rcx)  ; loop start
5555555A7E56: 0F 84 56 02 00 00          je     0x5555555a80b2  ; <+802>
5555555A7E5C: 49 8B 34 0E                movq   (%r14,%rcx), %rsi
5555555A7E60: 8B 3E                      movl   (%rsi), %edi
5555555A7E62: 3B 3A                      cmpl   (%rdx), %edi
5555555A7E64: 48 0F 4E F2                cmovleq %rdx, %rsi
5555555A7E68: 48 0F 45 D6                cmovneq %rsi, %rdx
5555555A7E6C: 0F B6 C0                   movzbl %al, %eax
5555555A7E6F: 41 0F 45 C7                cmovnel %r15d, %eax
5555555A7E73: 48 83 C1 10                addq   $0x10, %rcx
5555555A7E77: 49 39 CC                   cmpq   %rcx, %r12
5555555A7E7A: 75 D4                      jne    0x5555555a7e50  ; <+192>  loop end
5555555A7E7C: CC                         int3   
```

Reply:
Compiler is being dumb around the choice of size for bool though, causing it to generate the unnecessary `movzbl` instruction in both versions. If we change the is_equal variable to be a usize, we can get rid of that. This, of course, does not improve the readability of the code.
```rust
#[inline]
fn test_equality<'a, T: Ord>(slices: &[&'a [T]]) -> Equality<'a, T> {
    let mut is_equal: usize = 1;
    let mut max = &slices[0][0];
    unsafe {std::arch::asm!("int3");}
    for s in slices {
        let x = &s[0];
        if x != max {
            is_equal = 0;
            if x > max {
                max = x;
            }
        }
    }
    unsafe {std::arch::asm!("int3");}
    if is_equal != 0 { Equal(max) } else { NotEqual(max) }
}
```
Assembly:
```assembly
// 11 instructions in the loop - no more movzbl
5555555A7E3E: CC                         int3   
5555555A7E3F: B8 01 00 00 00             movl   $0x1, %eax
5555555A7E44: 31 C9                      xorl   %ecx, %ecx
5555555A7E46: 66 2E 0F 1F 84 00 00 00 >  nopw   %cs:(%rax,%rax)
5555555A7E50: 49 83 7C 0E 08 00          cmpq   $0x0, 0x8(%r14,%rcx)  ; loop start
5555555A7E56: 0F 84 46 02 00 00          je     0x5555555a80a2  ; <+786>
5555555A7E5C: 49 8B 34 0E                movq   (%r14,%rcx), %rsi
5555555A7E60: 8B 3E                      movl   (%rsi), %edi
5555555A7E62: 3B 3A                      cmpl   (%rdx), %edi
5555555A7E64: 48 0F 4E F2                cmovleq %rdx, %rsi
5555555A7E68: 48 0F 45 D6                cmovneq %rsi, %rdx
5555555A7E6C: 49 0F 45 C7                cmovneq %r15, %rax
5555555A7E70: 48 83 C1 10                addq   $0x10, %rcx
5555555A7E74: 49 39 CC                   cmpq   %rcx, %r12
5555555A7E77: 75 D7                      jne    0x5555555a7e50  ; <+192>  loop end
5555555A7E79: CC                         int3   
```

Reply:
Anyway, I don't think this will make a significant difference in performance either way. Just let me know what you feel is best and I'll go with that.